src/lib/support/BufferWriter.h

```diff
@@ -0,0 +1,194 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <climits>
+#include <stdint.h>
+#include <string.h>
+
+namespace chip {
+namespace Encoding {
+
+template <class Derived>
+class BufferWriterBase
+{
+public:
+    /// Append a null terminated string, exclude the null terminator
+    /// This does NOT care about endianess
+    BufferWriterBase & Put(const char * s)
+    {
+        static_assert(CHAR_BIT == 8, "We're assuming char and uint8_t are the same size");
+        while (*s != 0)
+        {
+            Put8(static_cast<uint8_t>(*s++));
+        }
+        return *this;
+    }
+
+    /// Raw append a buffer, regardless of endianess
+    BufferWriterBase & Put(const void * buf, size_t len)
+    {
+        size_t available = Available();
+
+        memmove(mBuf + mNeeded, buf, available < len ? available : len);
```

Comment:
If `available == 0`, `mBuf + mNeeded` might not be a valid pointer at all, in which case I think this is technically undefined behavior per spec...